### ISSUE TITLE
[BUG] PPO training has 5 critical defects — policy cannot learn effectively

### ISSUE THREAD

**Issue by winktool** [lubluniky/rocket-rs#24]
### Summary

A deep audit of the full training and inference pipeline across `src/algorithms/ppo.rs`, `src/algorithms/rollout.rs`, and `src/simd/gae.rs` revealed **5 critical bugs**, 2 of which are fatal (the policy fundamentally cannot learn). Additionally, there are 5 medium-severity issues and 2 well-implemented modules (`ActorCritic`, `CheckpointManager`) that PPO completely ignores.

### Environment

- **OS**: macOS (Apple Silicon / aarch64)
- **Rust**: edition 2024
- **candle**: candle-core + candle-nn
- **Test status**: 50 quant + 9 envs tests pass (unit tests do not cover training convergence)

---

### BUG 1: 🔴 `log_std` not registered in VarMap — policy std is frozen at 1.0 forever

**Location**: `src/algorithms/ppo.rs` L160

```rust
self.log_std = Some(Tensor::zeros(&[self.act_dim], DType::F32, &candle_device)?);
```

`log_std` is a plain `Tensor::zeros`, **not in `VarMap`**. Since `AdamW::new(self.var_map.all_vars(), ...)` only updates variables inside VarMap, `log_std` is **never updated** — it stays at 0 forever, meaning `std = exp(0) = 1.0`.

**Impact**:
- Exploration noise in continuous action space is constant and uncontrollable
- The policy can never converge regardless of training duration
- This is a **root cause** of PPO training being ineffective

**Expected behavior**: `log_std` should be registered in VarMap via `VarBuilder::get_with_hints()` so the optimizer can update it.

---

### BUG 2: 🔴 Network capacity severely insufficient — [64, 64] for 662-dim input

**Location**: `src/algorithms/ppo.rs` L85

```rust
let hidden_sizes = vec![64, 64];
```

`TradingEnv` observation dim = `lookback(20) × features(33) + 2 = 662`. A 2-layer MLP with only 64 hidden units creates a severe information bottleneck.

**Comparison**: The same codebase's `ActorCriticConfig` defaults to `[256, 256]` with LSTM/GRU support and orthogonal initialization — but PPO does not use it at all.

**Impact**: Insufficient network expressiveness. Even if all other bugs were fixed, the network cannot learn an effective policy.

---

### BUG 3: 🟠 AdamW optimizer recreated on every `update()` call — momentum completely lost

**Location**: `src/algorithms/ppo.rs` L413

```rust
fn update(&mut self, buffer: &RolloutBuffer) -> Result<TrainMetrics> {
    // ...
    let mut optimizer = AdamW::new(self.var_map.all_vars(), params)?;
    // ...
}
```

A new `AdamW` is created on every `collect_rollout() → update()` cycle. This resets the first-moment (m) and second-moment (v) exponential moving averages to zero, effectively degrading Adam to SGD with decay.

**Impact**: Unstable learning; the core adaptive learning rate benefit of Adam is entirely lost.

**Expected behavior**: The optimizer should be a persistent field of `PPOAgent`, created once. Learning rate updates should use `set_learning_rate()`.

---

### BUG 4: 🟠 Gradient clipping never implemented — `max_grad_norm` defined in config but unused

**Location**: `src/algorithms/config.rs` L46 defines `max_grad_norm: f32` (default 0.5), but `ppo.rs` `update()` only calls `optimizer.backward_step(&total_loss)` with no gradient clipping.

```rust
// config.rs
pub max_grad_norm: f32, // default 0.5 ← defined

// ppo.rs update()
optimizer.backward_step(&total_loss); // ← direct backward+step, no grad clip
```

**Impact**: Abnormal gradients cause parameter explosion, sporadic NaN, or sudden policy collapse. Standard PPO implementations (SB3, CleanRL) all use `max_grad_norm=0.5`.

---

### BUG 5: 🟠 SIMD GAE truncation handling is incorrect — always triggers on Apple Silicon

**Location**: `src/algorithms/rollout.rs` L270-295, `src/simd/gae.rs` NEON kernel

```rust
// rollout.rs: merges terminated and truncated into a single done mask
let dones: Vec<f32> = self.terminated.iter()
    .zip(self.truncated.iter())
    .map(|(&t, &tr)| t.max(tr))
    .collect();

// simd/gae.rs NEON: zeros next_value whenever done=1
let mask = vsubq_f32(one_vec, done);
let next_masked = vmulq_f32(next_value, mask); // V(s')=0 regardless of terminated/truncated
```

**Problem**: For **truncated** episodes (e.g., `episode_length` timeout), the correct behavior is to preserve the bootstrap value `V(s')`. The SIMD path conflates terminated and truncated, incorrectly zeroing `V(s')` for both.

**The scalar path is correct** (`compute_gae_scalar_with_truncation`), but on `aarch64+neon` (Apple Silicon), conditional compilation always hits the SIMD path which returns successfully — the scalar fallback never executes.

**Impact**: With default `episode_length=252`, nearly every episode ends via truncation → advantage estimates are systematically biased negative at episode boundaries → policy learning is distorted.

---

### Additional issues

| # | Severity | Description | Location |
|---|----------|-------------|----------|
| 6 | 🟡 | Missing value loss clipping | ppo.rs L465 |
| 7 | 🟡 | save/load omits log_std, optimizer state, and training progress | ppo.rs L716-747 |
| 8 | 🟡 | Reward is absolute price diff, not return rate — inconsistent scale across stocks | trading.rs L298 |
| 9 | 🟡 | `predict()` does not clamp continuous actions to [-1, 1] | ppo.rs L690 |
| 10 | 🟢 | No validation when batch_size > n_samples | ppo.rs L427 |

### Unused modules

- **`src/networks/actor_critic.rs`** (691 lines): Supports `[256,256]`, learnable log_std in VarMap, LSTM/GRU, orthogonal init, shared backbone — **completely unused by PPO**
- **`src/checkpoint.rs`**: Atomic saves, auto cleanup, best model tracking, optimizer state persistence — **completely unused by PPO**

### Suggested fix path

1. **P0**: Migrate PPO to the existing `ActorCritic` network module → fixes BUG 1 + BUG 2 in one shot
2. **P1**: Promote optimizer to a persistent `PPOAgent` field → fixes BUG 3
3. **P1**: Implement `backward()` + `clip_grad_norm()` + `step()` → fixes BUG 4
4. **P1**: Split SIMD GAE interface into separate `terminated` / `dones` masks → fixes BUG 5
5. **P2**: Migrate to `CheckpointManager`, fix save/load
6. **P2**: Address additional issues 6-10

**Comment by imkow** [comment]
@winktool Someone depends on coding AI too heavily, don't take it too seriously.

**Comment by lubluniky** [comment]
Fixed by PR #29 with targeted PPO/GAE stability changes and regression tests.\n\nCoverage references:\n- src/algorithms/ppo.rs (space detection, optimizer/log_std/grad clipping related behavior)\n- src/algorithms/rollout.rs + src/simd/gae.rs (terminated vs truncated bootstrap semantics)\n- src/buffer/nstep.rs, src/checkpoint.rs, src/core/precision.rs (related regressions)\n\nFollow-up dependency migrations are tracked separately in #30, #31, #32, #33, #34.
